2단계 - 메모 (Create, Read)

BeokBeok opened the pull request:
# 작업 내용
- 리스트 형태의 메모 앱 구현
- + 버튼을 누르면 화면에 텍스트를 입력할 수 있는 추가 화면 노출
- 추가 화면에는 텍스트 입력 칸, 취소 버튼, 확인 버튼 있음
- 추가 화면에서 텍스트를 입력하고 확인 버튼을 누르면 메모가 저장되고 리스트 화면에서는 저장된 메모가 업데이트됨

BeokBeok committed:
:sparkles: feat: Memo 클래스 추가

BeokBeok committed:
:sparkles: feat: MemoService 구현

BeokBeok committed:
:sparkles: feat: fetch 함수 구현

BeokBeok committed:
:sparkles: feat: MemoRepository 추가

BeokBeok committed:
:sparkles: feat: MemoRepositoryImpl 구현

BeokBeok committed:
:sparkles: feat: domain 모듈 내 repository 패키지 생성

BeokBeok committed:
:sparkles: feat: 코루틴 디펜던시 추가

BeokBeok committed:
:sparkles: feat: mockk 디펜던시 추가

BeokBeok committed:
:sparkles: feat: MainViewModel 구현

BeokBeok committed:
:sparkles: feat: 데이터바인딩 적용

BeokBeok committed:
:sparkles: feat: fragment-ktx 디펜던시 추가

BeokBeok committed:
:sparkles: feat: MainAdapter 및 ViewHolder

BeokBeok committed:
:sparkles: feat: RecyclerView 설정

BeokBeok committed:
:sparkles: feat: 메모 불러오기 기능 구현

BeokBeok committed:
:sparkles: feat: MemoWriteViewModel 구현

BeokBeok committed:
:sparkles: feat: MemoWriteActivity 구현

BeokBeok committed:
:sparkles: feat: 메모 작성 버튼 클릭 시 메모 작성 화면 이동

BeokBeok committed:
:sparkles: feat: 메모 작성 화면에서 확인, 취소 버튼 이벤트 구현

BeokBeok committed:
:sparkles: feat: MemoService 및 MemoREpositoryImpl 싱글턴 적용

BeokBeok committed:
:white_check_mark: test: 빈 메모를 작성하고 저장 버튼을 누르면 메모가 저장되지 않습니다 테스트 함수

BeokBeok committed:
:sparkles: feat: main 패키지 생성 및 이동

BeokBeok committed:
:sparkles: feat: write 패키지 생성 및 이동

galcyurio commented on app/src/main/java/camp/nextstep/edu/memo/main/MainViewModel.kt:
```diff
@@ -0,0 +1,20 @@
+) : ViewModel() {
+
+    private val _memoList = MutableStateFlow<List<Memo>>(emptyList())
+    val memoList: StateFlow<List<Memo>> get() = _memoList
```
실제로는 MutableStateFlow가 반환되므로 `_memoList.asStateFlow()`를 통해 ReadOnly 구현체를 반환해주세요.

BeokBeok replied:
asStateFlow() 를 쓰면 되는군요!
하나 배워갑니다 😄 
[79b1ea8]

galcyurio commented on app/src/main/java/camp/nextstep/edu/memo/write/MemoWriteActivity.kt:
```diff
@@ -0,0 +1,60 @@
+                            finish()
+                        }
+                        else -> Unit
+                    }
```
sealed class 또는 enum class와 함께 쓰는 경우에는 when expression과 함께 쓰면 else 문을 생략할 수 있습니다.

아래 글에서 설명하는 여러 방법들 중 하나를 적용해보세요.
https://github.com/cashapp/exhaustive#alternatives-considered

BeokBeok replied:
적용하였습니다
[bf7c088]

한가지 질문이 있는데요
만약 아래와 같이 정의되어있는 경우
```
sealed class Alphabet {
    object A : Alphabet()
    object B : Alphabet()
    object C : Alphabet()
    object D : Alphabet()
}
```
이 중에서 A 와 B 만 필요할 경우에도 else 를 사용하지 않고 C, D 모두 풀어서 써주는 것이 좋나요? 🤔 
가령
```
when (it) {
    A -> {}
    B -> {}
    C,
    D -> {}
}
```
와 
```
when (it) {
    A -> {}
    B -> {}
    else -> {}
}

```
중에서 전자가 더 좋나요?

galcyurio replied:
개발자에 따라 다르겠지만 제 경우에는 전자를 더 선호합니다.
그 이유는 전자와 같이 코드를 작성하면 추후에 Alphabet sealed class에 `object E : Alphabet()`에 추가되었을 때 when expression에서 컴파일 오류를 발생시키기 때문입니다.

이건 sealed class만 그런게 아니라 합타입(sealed class, sealed interface, enum class) 은 모두 해당됩니다.
조금 더 자세한 내용은 아래 블로그 글을 참조해보세요.
https://wooooooak.github.io/kotlin/2020/06/08/Kolint_exhastive/

galcyurio commented on app/src/main/java/camp/nextstep/edu/memo/write/MemoWriteActivity.kt:
```diff
@@ -0,0 +1,60 @@
+
+    private fun setupObserver() {
+        lifecycleScope.launch {
+            repeatOnLifecycle(Lifecycle.State.STARTED) {
```
이런 패턴은 많이 반복될 것 같은데 별도의 함수로 분리해보는 건 어떨까요?

BeokBeok replied:
분리해보았습니다
[304ce88]

galcyurio commented on app/src/main/java/camp/nextstep/edu/memo/write/MemoWriteViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+import kotlinx.coroutines.flow.StateFlow
+
+class MemoWriteViewModel(
+    private val memoRepository: MemoRepository = MemoRepositoryImpl.getInstance()
```
생성자 주입 👍

galcyurio commented on app/src/main/res/layout/activity_memo_write.xml:
```diff
@@ -0,0 +1,48 @@
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="20dp"
+            android:onClick="@{() -> viewModel.saveMemo(``)}"
```
별도로 cancel() 함수를 만드는게 더 명확할 것 같아요.

BeokBeok replied:
cancel 함수 구현하였습니다
[6299e7b]

galcyurio commented on app/src/main/res/layout/activity_memo_write.xml:
```diff
@@ -0,0 +1,48 @@
+            android:id="@+id/btn_memo_write_confirm"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:onClick="@{() -> viewModel.saveMemo(etMemoWrite.getText().toString())}"
```
`et_memo_write` EditText에 대해 양방향 데이터 바인딩을 걸고 ViewModel에서 처리하면 단위 테스트에 더 용이하고 xml에서 로직을 분리할 수 있을 것 같아요.

BeokBeok replied:
버튼을 클릭 했을 때에만 EditText에 입력된 데이터를 받아서 처리하기 때문에, 굳이 양방향 데이터 바인딩을 걸 필요가 없다고 생각했습니다.

양방향 데이터 바인딩을 걸게 되면 MutableLiveData가 public으로 열리게 되서, 양방향 데이터 바인딩이 반드시 필요할 경우에만 적용하고 있는데요.
가령, 패스워드를 8글자 이상 입력해야만 하는 경우와 같이, 뷰에서 입력한 데이터를 실시간으로 뷰모델에서 체크해서 8글자 미만일 때 경고 메시지를 노출이 필요하면 양방향 데이터 바인딩을 사용합니다.

단위테스트도 saveMemo 함수 파라미터로 값을 받을 수 있기 때문에 차이는 없을 것 같다고 생각하고, xml에서 로직을 분리하는 부분에 대해서도 어차피 xml에서 onClick을 사용하는 것은 동일하지 않나요? 🤔 

태현님 의견은 어떠신지 궁금합니다 🙏

galcyurio replied:
말씀하신 부분처럼 굳이 분리하지 않아도 단위테스트에는 크게 차이가 없을 것 같습니다.
저는 View에서 최대한 코드를 쓰지않는다는 생각을 지니고 있어서 위와 같은 리뷰를 드렸습니다.
만약 입력한 값이 비어있으면 버튼을 비활성화하는 등 새로운 기능이 추가되면 분리했을 때 이점이 있겠지만 현재의 요구사항에서는 굳이 분리할 필요는 없어보이네요.

galcyurio commented on app/src/test/java/camp/nextstep/edu/memo/main/MainViewModelTest.kt:
```diff
@@ -0,0 +1,37 @@
+
+internal class MainViewModelTest {
+
+    private val memoRepository: MemoRepository = mockk(relaxed = true)
```
Mock 객체 활용 👍

galcyurio commented on app/src/main/java/camp/nextstep/edu/memo/write/MemoWriteViewModel.kt:
```diff
@@ -0,0 +1,26 @@
+) : ViewModel() {
+
+    private val _isSavedMemo = MutableStateFlow<MemoEvent>(MemoEvent.None)
+    val isSavedMemo: StateFlow<MemoEvent> get() = _isSavedMemo
```
위와 마찬가지로 `isSavedMemo.asStateFlow()` 호출 해주세요.

BeokBeok replied:
적용하였습니다.
[7f12103]

galcyurio replied:
이 부분 제대로 안보고 지나갔었네요.
`isSavedMemo` 보다는 `MemoEvent`에 어울리는 네이밍을 해주면 좋겠어요.

galcyurio replied:
Event를 StateFlow로 처리하는 경우 이슈가 발생할 수 있습니다.
저도 Flow에서 이벤트 처리하는 방식은 요즘 많이 고민하고 있어서 어떤 방식이 괜찮은지 판단이 서지 않아 명확하게 피드백을 드리기 힘드네요. 죄송합니다.

시간 나실 때 아래 글을 읽어보면 좋겠습니다.

https://elizarov.medium.com/shared-flows-broadcast-channels-899b675e805c
https://proandroiddev.com/android-singleliveevent-redux-with-kotlin-flow-b755c70bb055

BeokBeok replied:
memoEvent 로 변수명을 변경해보았습니다 😄 
[f53e9d343a2db6502294745a4a35c4a5b299d454]

BeokBeok replied:
저도 StateFlow를 SwipeRefreshLayout에 사용했었다가 이슈가 발생해서 LiveData로 변경했던 경험이 있네요 😭 
StateFlow 내부 코드를 보니, 이전값과 새로운값이 다를 때에만 collect가 발생한다는 것을 알게되었습니다 😓 
이미 알고 계실수도 있으시겠지만, 생각나서 공유드립니다 💦 
```
    override suspend fun collect(collector: FlowCollector<T>) {
            /* 중략 */
            while (true) {
                /* 중략 */
                if (oldState == null || oldState != newState) {
                    collector.emit(NULL.unbox(newState))
                    oldState = newState
                }
               /* 중략 */
    }
```

galcyurio commented on data/src/test/java/camp/nextstep/edu/memo/data/MemoServiceTest.kt:
```diff
@@ -0,0 +1,37 @@
+        assertThat(memoService.memoList.first()).isEqualTo(memo)
+    }
+
+    @Test
+    fun `작성된 메모들을 불러옵니다`() {
+        memoService = mockk()
+        val memoList = listOf(
+            Memo("메모1"),
+            Memo("메모2")
+        )
+
+        every {
+            memoService.fetch()
+        } returns memoList
+
+        assertThat(memoService.fetch()).isEqualTo(memoList)
+    }
```
Mock 객체의 fetch()에서 memoList 반환하도록 stubbing 하고 fetch()를 호출한 뒤 memoList가 나왔는지를 검사하고 있네요.

MemoService를 인스턴스화하는 대신에 실수로 mock 객체로 테스트하는 코드를 만들어버리신 것 같아요.

BeokBeok replied:
memoService를 인스턴스화해서 테스트하니까 every, verify 등 모든 mockk 함수 사용 시 에러가 발생하더라구요 😭 

혹시 인스턴스화하고도 동작이 되는 방법을 아시나요!? 🙏

olaf-prnd replied:
`spyk(instance)`를 활용하시면 되지만 지금은 Spy 객체를 쓸 필요가 없습니다.

현재 `Stubbing MemoService.fetch() -> Assert MemoService.fetch()`를 하고 있어서 MemoService.fetch() 자리에 그 어떤 클래스의 함수가 들어가든 성공하는 테스트입니다.

아래와 같이 public으로 공개된 함수들을 이용해서 MemoService에 미리 Memo를 추가해주세요.
```kotlin
    @Test
    fun `작성된 메모들을 불러옵니다`() {
        // given: '메모1', '메모2'를 저장한다.
        memoService = MemoService.getInstance()
        memoService.save(Memo("메모1"))
        memoService.save(Memo("메모2"))

        // when: 저장된 메모들을 불러온다.
        val actual = memoService.fetch()

        // then: 저장했던 '메모1', '메모2'가 포함되어야 한다.
        assertThat(actual).contains(Memo("메모1"), Memo("메모2"))
    }
```

galcyurio replied:
`spyk(instance)`를 활용하시면 되지만 지금은 Spy 객체를 쓸 필요가 없습니다.

현재 `Stubbing MemoService.fetch() -> Assert MemoService.fetch()`를 하고 있어서 MemoService.fetch() 자리에 그 어떤 클래스의 함수가 들어가든 성공하는 테스트입니다.

아래와 같이 public으로 공개된 함수들을 이용해서 MemoService에 미리 Memo를 추가해주세요.
```kotlin
    @Test
    fun `작성된 메모들을 불러옵니다`() {
        // given: '메모1', '메모2'를 저장한다.
        memoService = MemoService.getInstance()
        memoService.save(Memo("메모1"))
        memoService.save(Memo("메모2"))

        // when: 저장된 메모들을 불러온다.
        val actual = memoService.fetch()

        // then: 저장했던 '메모1', '메모2'가 포함되어야 한다.
        assertThat(actual).contains(Memo("메모1"), Memo("메모2"))
    }
```

BeokBeok replied:
아직 테스트코드에 대한 지식이 부족한 것 같습니다 😭 
fetch 함수에 대한 테스트를 작성한다면 웬지 부가적인 함수(예, save함수)를 호출해서 조건을 만들기보다는 mocking 해서 조건을 만들어서 테스트 하는 것에 익숙해서 그랬던 것 같습니다!
생각해보니 mocking 이나 부가적인 함수를 호출하여 조건을 만드는거나 동일하기 때문에, 적어주신 대로 작성하는 것이 좋겠네요! 👍 
감사합니다!
[8a0a683d1ad2bc01cfeeff3e8ea953ca5dde1c28]

galcyurio commented on data/src/test/java/camp/nextstep/edu/memo/data/MemoRepositoryImplTest.kt:
```diff
@@ -0,0 +1,43 @@
+        } returns memoList
+
+        Assertions.assertThat(memoService.fetch()).isEqualTo(memoList)
+    }
```
이 테스트도 조금 이상하네요.
MemoService Mock 객체에 대해 `fetch() -> memoList`와 같이 stubbing을 한뒤 `fetch() -> memoList`를 체크하고 있습니다.
repository.fetch()를 잘못 호출하신 것 같아요.

BeokBeok replied:
앗! 실수했네요 😭 
수정하였습니다 😅 
[b268695]

galcyurio commented on data/src/test/java/camp/nextstep/edu/memo/data/MemoRepositoryImplTest.kt:
```diff
@@ -0,0 +1,43 @@
+
+    @Test
+    fun `작성한 메모를 저장합니다`() {
+        repository = MemoRepositoryImpl(service = memoService)
```
MemoRepository 생성이 반복되는데 `@BeforeEach`로 한 번만 해주면 어떨까요?

```Kotlin
private lateinit var memoService: MemoService
private lateinit var repository: MemoRepository

@BeforeEach
fun setUp() {
    memoService = mockk(relaxed = true)
    repository = MemoRepositoryImpl(service = memoService)
}
```

아래와 같이 MockK의 annotation을 활용하는 방식도 있습니다.

```kotlin
@MockK(relaxed = true)
lateinit var memoService: MemoService
private lateinit var repository: MemoRepository

@BeforeEach
fun setUp() {
    MockKAnnotations.init(this)
    repository = MemoRepositoryImpl(memoService)
}
```

BeokBeok replied:
메모 생성하는 테스트가 메모를 불러오는 테스트에 영향을 미칠까봐 매번 생성을 하였는데요
생각해보니 MemoService를 mockking 해서 사용하니 괜찮을 것 같네요 😄 
[a4fa369]

olaf-prnd replied:
`@BeforeEach`로 선언해둔 함수는 각 `@Test` 함수 실행전에 실행됩니다.
따라서 각 `@Test` 함수는 이전 테스트에 영향을 받지 않는 새로운 `MemoRepository`로 테스트하게 됩니다.

galcyurio replied:
`@BeforeEach`로 선언해둔 함수는 각 `@Test` 함수 실행전에 실행됩니다.
따라서 각 `@Test` 함수는 이전 테스트에 영향을 받지 않는 새로운 `MemoRepository`로 테스트하게 됩니다.

galcyurio commented on data/src/main/java/camp/nextstep/edu/memo/data/MemoService.kt:
```diff
@@ -0,0 +1,26 @@
+class MemoService {
+
+    private val _memoList = mutableListOf<Memo>()
+    val memoList: List<Memo> get() = _memoList
```
이 필드는 테스트 코드에서만 쓰이고 있습니다. 제거하고 프로덕션 코드에서 쓰이고 있는 fetch()를 통해 테스트 해주세요.

BeokBeok replied:
수정하였습니다
[d993d76]

galcyurio commented on data/src/main/java/camp/nextstep/edu/memo/data/MemoRepositoryImpl.kt:
```diff
@@ -0,0 +1,26 @@
+    override fun fetch(): List<Memo> = service.fetch()
+
+    companion object {
+        @Volatile
+        private var instance: MemoRepository? = null
+
+        fun getInstance(): MemoRepository = instance ?: synchronized(this) {
+            instance ?: MemoRepositoryImpl(service = MemoService.getInstance()).also {
+                instance = it
+            }
+        }
```
싱글턴 패턴 👍

galcyurio reviewed (changes requested):
의존성 주입과 Event sealed class 사용까지 👍 👍 
몇 가지 개선할 점이 보여서 코멘트 남겨두었습니다.
피드백 반영 후 다시 요청해주세요!

BeokBeok committed:
:art: review: 실제로는 MutableStateFlow가 반환되므로 _memoList.asStateFlow()를 통해 ReadOnly 구현체를 반환해주세요.

BeokBeok committed:
:art: review: sealed class 또는 enum class와 함께 쓰는 경우에는 when expression과 함께 쓰면 else 문을 생략할 수 있습니다.

              아래 글에서 설명하는 여러 방법들 중 하나를 적용해보세요.
              https://github.com/cashapp/exhaustive#alternatives-considered

BeokBeok committed:
:art: review: 이런 패턴은 많이 반복될 것 같은데 별도의 함수로 분리해보는 건 어떨까요?

BeokBeok committed:
:art: review: 별도로 cancel() 함수를 만드는게 더 명확할 것 같아요.

BeokBeok committed:
:art: review: 위와 마찬가지로 isSavedMemo.asStateFlow() 호출 해주세요.

BeokBeok committed:
:art: review: 이 테스트도 조금 이상하네요.
              MemoService Mock 객체에 대해 fetch() -> memoList와 같이 stubbing을 한뒤 fetch() -> memoList를 체크하고 있습니다.
              repository.fetch()를 잘못 호출하신 것 같아요.

BeokBeok committed:
:art: review: MemoRepository 생성이 반복되는데 @BeforeEach로 한 번만 해주면 어떨까요?

BeokBeok committed:
:art: review: 이 필드는 테스트 코드에서만 쓰이고 있습니다. 제거하고 프로덕션 코드에서 쓰이고 있는 fetch()를 통해 테스트 해주세요.

olaf-prnd reviewed (approved):
피드백 잘 반영해주셨네요 👍 👍 
질문해주신 부분들은 쓰레드 이어서 코멘트 남겨두었습니다.
추가로 몇 가지 피드백을 남겨두었는데 수정사항이 크지 않아 바로 병합하겠습니다.
다음 단계에 같이 반영해주세요!

galcyurio reviewed (approved):
피드백 잘 반영해주셨네요 👍 👍 
질문해주신 부분들은 쓰레드 이어서 코멘트 남겨두었습니다.
추가로 몇 가지 피드백을 남겨두었는데 수정사항이 크지 않아 바로 병합하겠습니다.
다음 단계에 같이 반영해주세요!
